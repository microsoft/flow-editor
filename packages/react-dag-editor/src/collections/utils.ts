export const is = Object.is;

export class MapIterator<T, U> implements IterableIterator<U> {
  private readonly upstream: Iterator<T>;
  private readonly f: (value: T) => U;

  public constructor(upstream: Iterator<T>, f: (value: T) => U) {
    this.upstream = upstream;
    this.f = f;
  }

  public [Symbol.iterator](): IterableIterator<U> {
    return this;
  }

  public next(): IteratorResult<U> {
    const next = this.upstream.next();
    if (next.done) {
      return next;
    }
    return {
      done: false,
      value: this.f(next.value)
    };
  }
}
