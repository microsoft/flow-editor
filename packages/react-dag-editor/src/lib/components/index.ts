export * from "./ReactDagEditor";
export * from "./Graph/Graph";
export * from "./ItemPanel";
export * from "./Minimap/Minimap";
export * from "./StaticGraph/StaticGraph";
export * from "./Graph/IGraphProps";
export * from "./RegisterComponent";
export { defaultGetPositionFromEvent } from "../controllers";
export * from "./GridBackground";
export * from "./NodeAnchors";
