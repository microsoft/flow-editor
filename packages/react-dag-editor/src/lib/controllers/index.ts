export * from "./DragController";
export * from "./DragNodeController";
export * from "./TouchController";
export * from "./TwoFingerHandler";
