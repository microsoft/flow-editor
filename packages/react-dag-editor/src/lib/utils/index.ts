export * from "./getPortPosition";
export * from "./getNodeConfig";
export * from "./graphDataUtils";
export * from "./transformMatrix";
export * from "./getCurvePathD";
export * from "./layout";
export * from "./browsers";
export * from "./isWithinThreshold";
export * from "./geometric";
export * from "./viewport";
export * from "./Writable";
export * from "./gantt/getCriticalPath";
export * from "./gantt/getTopoSortingNodes";
export * from "./getOffsetLimit";
export * from "./event";
export * from "./valueOrCallback";
export * from "./debug";
export * from "./debounce";
export * from "./arrayEquals";
export * from "./empty";
export * from "./stopPropagation";
export * from "./clamp";
export * from "./history";
export * as Bitset from "./bitset";
