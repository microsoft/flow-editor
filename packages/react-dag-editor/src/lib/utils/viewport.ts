import type * as React from "react";
import type { HashMap, OrderedMap } from "../collections";
import type { IGraphConfig } from "../models/config/types";
import type { ICanvasEdge } from "../models/edge";
import type { EdgeModel } from "../models/EdgeModel";
import {
  Direction,
  IContainerRect,
  IGap,
  IPoint,
  IRectShape,
  IRectSize,
  ITransformMatrix,
  IViewport,
} from "../models/geometry";
import type { GraphModel } from "../models/GraphModel";
import type { ICanvasNode } from "../models/node";
import type { NodeModel } from "../models/NodeModel";
import type { IGraphSettings } from "../models/state";
import { isPointInRect } from "./geometric";
import { identical } from "./identical";
import { getNodeSize, IShapeRect } from "./layout";
import { getContainerClientPoint, reverseTransformPoint } from "./transformMatrix";

export const isViewportComplete = (viewport: IViewport): viewport is Required<IViewport> => {
  return !!viewport.rect;
};

export const isViewportEmpty = (viewport: IViewport): boolean => {
  if (!isViewportComplete(viewport)) {
    return true;
  }
  return !viewport.rect.width || !viewport.rect.height;
};

export const getNodeRect = (node: ICanvasNode, graphConfig: IGraphConfig): IShapeRect => {
  const { x, y } = node;
  const { width, height } = getNodeSize(node, graphConfig);
  return {
    x,
    y,
    width,
    height,
  };
};

export const isNodeVisible = (node: ICanvasNode, viewport: Required<IViewport>, graphConfig: IGraphConfig): boolean => {
  return isRectVisible(getNodeRect(node, graphConfig), viewport);
};

export const isRectVisible = (shapeRect: IShapeRect, viewport: Required<IViewport>) => {
  const { x, y, width, height } = shapeRect;
  return (
    isPointVisible({ x, y }, viewport) ||
    isPointVisible({ x: x + width, y }, viewport) ||
    isPointVisible({ x: x + width, y: y + height }, viewport) ||
    isPointVisible({ x, y: y + height }, viewport)
  );
};

export const isPointVisible = (point: IPoint, viewport: Required<IViewport>): boolean => {
  const { x, y } = getContainerClientPoint(point.x, point.y, viewport);

  const { height, width } = viewport.rect;

  return x > 0 && x < width && y > 0 && y < height;
};

export const getVisibleNodes = <NodeData, PortData>(
  nodes: OrderedMap<string, NodeModel<NodeData, PortData>>,
  viewport: Required<IViewport>,
  graphConfig: IGraphConfig,
): ICanvasNode[] => {
  const result: ICanvasNode[] = [];

  nodes.forEach(n => {
    if (isNodeVisible(n, viewport, graphConfig)) {
      result.push(n.inner);
    }
  });

  return result;
};

// Get rendered nodes count
export const getRenderedNodes = <NodeData, PortData>(
  nodes: OrderedMap<string, NodeModel<NodeData, PortData>>,
  viewport: IViewport,
): ICanvasNode[] => {
  const result: ICanvasNode[] = [];

  const renderedArea = getRenderedArea(viewport);

  nodes.forEach(n => {
    if (isNodeInRenderedArea(n, renderedArea)) {
      result.push(n.inner);
    }
  });

  return result;
};

const isNodeInRenderedArea = (node: ICanvasNode, renderedArea: IRectShape): boolean => {
  return isPointInRect(renderedArea, node);
};

// Get rendered edges count
export const getRenderedEdges = (
  edges: HashMap<string, EdgeModel>,
  nodes: OrderedMap<string, NodeModel>,
  graphConfig: IGraphConfig,
  viewport: IViewport,
) => {
  const result: ICanvasEdge[] = [];
  const renderedArea = getRenderedArea(viewport);

  edges.forEach(e => {
    const edgeCoordinate = getEdgeSourceTargetCoordinate(e, nodes, graphConfig);
    if (
      edgeCoordinate.source &&
      edgeCoordinate.target &&
      isEdgeInRenderedArea(edgeCoordinate.source, edgeCoordinate.target, renderedArea)
    ) {
      result.push(e.inner);
    }
  });

  return result;
};

const isEdgeInRenderedArea = (source: IPoint, target: IPoint, renderedArea: IRectShape): boolean => {
  const isSourceVisible = isPointInRect(renderedArea, source);
  const isTargetVisible = isPointInRect(renderedArea, target);

  return isSourceVisible || isTargetVisible;
};

export const getVisibleArea = (viewport: IViewport): IRectShape => {
  if (!isViewportComplete(viewport)) {
    return {
      minX: 0,
      minY: 0,
      maxX: 0,
      maxY: 0,
    };
  }

  const { rect, transformMatrix } = viewport;
  const minX = 0;
  const minY = 0;
  const maxX = rect.width;
  const maxY = rect.height;
  const min = reverseTransformPoint(minX, minY, transformMatrix);
  const max = reverseTransformPoint(maxX, maxY, transformMatrix);

  return {
    minX: min.x,
    minY: min.y,
    maxX: max.x,
    maxY: max.y,
  };
};

export const getRenderedArea = (viewport: IViewport): IRectShape => {
  if (!isViewportComplete(viewport)) {
    return {
      minX: 0,
      minY: 0,
      maxX: 0,
      maxY: 0,
    };
  }

  const { rect, transformMatrix } = viewport;
  const minX = 0;
  const minY = 0;
  const maxX = rect.width;
  const maxY = rect.height;
  const min = reverseTransformPoint(minX - rect.width, minY - rect.height, transformMatrix);
  const max = reverseTransformPoint(maxX + rect.width, maxY + rect.height, transformMatrix);

  return {
    minX: min.x,
    minY: min.y,
    maxX: max.x,
    maxY: max.y,
  };
};

export const getEdgeSourceTargetCoordinate = (
  edge: ICanvasEdge,
  nodes: OrderedMap<string, NodeModel>,
  graphConfig: IGraphConfig,
): { source?: IPoint; target?: IPoint } => {
  return {
    source: nodes.get(edge.source)?.getPortPosition(edge.sourcePortId, graphConfig),
    target: nodes.get(edge.target)?.getPortPosition(edge.targetPortId, graphConfig),
  };
};

export interface IZoomCommonParams {
  nodeMinVisibleSize: IRectSize;
  nodeMaxVisibleSize: IRectSize;
}

export interface IZoomFixPublicParams {
  /**
   * zoom vertical
   */
  direction?: Direction;
  /**
   * is disable pan when zoom
   */
  disablePan?: boolean;
  spacing?: number | IGap;
}

const normalizeSpacing = (spacing: number | IGap | undefined): Required<IGap> => {
  if (!spacing) {
    return {
      top: 0,
      right: 0,
      bottom: 0,
      left: 0,
    };
  }
  if (typeof spacing === "number") {
    return {
      top: spacing,
      right: spacing,
      bottom: spacing,
      left: spacing,
    };
  }
  return {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    ...spacing,
  };
};

export interface IZoomFitParams extends IZoomFixPublicParams, IZoomCommonParams {
  data: GraphModel;
  graphConfig: IGraphConfig;
  rect: IContainerRect;
}

type Action = (viewport: IViewport) => IViewport;

/**
 * @param scale the scale to zoom to
 * @param anchor this client point is not moved after zoom
 * @param direction X: zoom horizontal, Y: zoom vertical
 */
export interface IZoomParams {
  scale: number;
  anchor: IPoint;
  direction?: Direction;
  limitScale(scale: number): number;
}

/**
 * zoom to [scale]
 */
export const zoomTo = ({ scale, anchor, direction, limitScale }: IZoomParams): Action => {
  return prevState => {
    const scaleX = limitScale(scale) / prevState.transformMatrix[0];
    const scaleY = limitScale(scale) / prevState.transformMatrix[3];
    const { x, y } = anchor;
    const dx = x * (1 - scaleX);
    const dy = y * (1 - scaleY);

    let transformMatrix: ITransformMatrix;
    switch (direction) {
      case Direction.X:
        transformMatrix = [
          scale,
          0,
          0,
          prevState.transformMatrix[3],
          prevState.transformMatrix[4] * scaleX + dx,
          prevState.transformMatrix[5],
        ];
        break;
      case Direction.Y:
        transformMatrix = [
          prevState.transformMatrix[0],
          0,
          0,
          scale,
          prevState.transformMatrix[4],
          prevState.transformMatrix[5] * scaleY + dy,
        ];
        break;
      case Direction.XY:
      default:
        transformMatrix = [
          scale,
          0,
          0,
          scale,
          prevState.transformMatrix[4] * scaleX + dx,
          prevState.transformMatrix[5] * scaleY + dy,
        ];
    }
    return {
      ...prevState,
      transformMatrix,
    };
  };
};

export const zoom = ({ scale, anchor, direction, limitScale }: IZoomParams): Action => {
  if (scale === 1) {
    return identical;
  }

  return prevState => {
    let transformMatrix: ITransformMatrix;
    switch (direction) {
      case Direction.X:
        return zoomTo({
          anchor,
          direction,
          limitScale,
          scale: prevState.transformMatrix[0] * scale,
        })(prevState);
      case Direction.Y:
        return zoomTo({
          anchor,
          direction,
          limitScale,
          scale: prevState.transformMatrix[3] * scale,
        })(prevState);
      case Direction.XY:
      default: {
        const resultX = limitScale(prevState.transformMatrix[0] * scale);
        const resultY = limitScale(prevState.transformMatrix[3] * scale);
        const scaleX = resultX / prevState.transformMatrix[0];
        const scaleY = resultY / prevState.transformMatrix[3];
        const { x, y } = anchor;
        const dx = x * (1 - scaleX);
        const dy = y * (1 - scaleY);
        transformMatrix = [
          resultX,
          0,
          0,
          resultY,
          prevState.transformMatrix[4] * scaleX + dx,
          prevState.transformMatrix[5] * scaleY + dy,
        ];
      }
    }
    return {
      ...prevState,
      transformMatrix,
    };
  };
};

export const pan = (dx: number, dy: number): Action => {
  if (dx === 0 && dy === 0) {
    return identical;
  }
  return prevState => {
    return {
      ...prevState,
      transformMatrix: [
        prevState.transformMatrix[0],
        prevState.transformMatrix[1],
        prevState.transformMatrix[2],
        prevState.transformMatrix[3],
        prevState.transformMatrix[4] + dx,
        prevState.transformMatrix[5] + dy,
      ],
    };
  };
};
export const minimapPan = (dx: number, dy: number): Action => {
  if (dx === 0 && dy === 0) {
    return identical;
  }
  return prevState => {
    const [a, b, c, d] = prevState.transformMatrix;
    return {
      ...prevState,
      transformMatrix: [
        a,
        b,
        c,
        d,
        prevState.transformMatrix[4] + a * dx + b * dy,
        prevState.transformMatrix[5] + c * dx + d * dy,
      ],
    };
  };
};

export const getContentArea = (data: GraphModel, graphConfig: IGraphConfig, nodeIds?: Set<string>) => {
  let minNodeWidth = Infinity;
  let minNodeHeight = Infinity;
  let minNodeX = Infinity;
  let minNodeY = Infinity;
  let maxNodeX = -Infinity;
  let maxNodeY = -Infinity;

  type Fn = (node: NodeModel) => void;

  const forEachNode = (fn: Fn) => data.nodes.forEach(fn);
  const forEachByIds = (fn: Fn) =>
    nodeIds?.forEach(id => {
      const node = data.nodes.get(id);
      if (node) {
        fn(node);
      }
    });

  const forEach = nodeIds === undefined ? forEachNode : forEachByIds;

  forEach(node => {
    const { width: nodeWidth, height: nodeHeight } = getNodeSize(node, graphConfig);

    if (node.x < minNodeX) {
      minNodeX = node.x;
    }
    if (node.y < minNodeY) {
      minNodeY = node.y;
    }
    if (node.x + nodeWidth > maxNodeX) {
      maxNodeX = node.x + nodeWidth;
    }
    if (node.y + nodeHeight > maxNodeY) {
      maxNodeY = node.y + nodeHeight;
    }
    if (nodeWidth < minNodeWidth) {
      minNodeWidth = nodeWidth;
    }
    if (nodeHeight < minNodeHeight) {
      minNodeHeight = nodeHeight;
    }
  });

  return {
    minNodeWidth,
    minNodeHeight,
    minNodeX,
    minNodeY,
    maxNodeX,
    maxNodeY,
  };
};

const normalizeNodeVisibleMinMax = ({ nodeMinVisibleSize, nodeMaxVisibleSize }: IZoomCommonParams) => {
  let { width: nodeMinVisibleWidth, height: nodeMinVisibleHeight } = nodeMinVisibleSize;
  let { width: nodeMaxVisibleWidth, height: nodeMaxVisibleHeight } = nodeMaxVisibleSize;
  if (nodeMinVisibleWidth > nodeMaxVisibleWidth) {
    const temp = nodeMinVisibleWidth;
    nodeMinVisibleWidth = nodeMaxVisibleWidth;
    nodeMaxVisibleWidth = temp;
  }
  if (nodeMinVisibleHeight > nodeMaxVisibleHeight) {
    const temp = nodeMinVisibleHeight;
    nodeMinVisibleHeight = nodeMaxVisibleHeight;
    nodeMaxVisibleHeight = temp;
  }
  return {
    nodeMinVisibleWidth,
    nodeMinVisibleHeight,
    nodeMaxVisibleWidth,
    nodeMaxVisibleHeight,
  };
};

export const getScaleRange = (params: IZoomCommonParams, { width: minNodeWidth, height: minNodeHeight }: IRectSize) => {
  const { nodeMinVisibleWidth, nodeMinVisibleHeight, nodeMaxVisibleWidth, nodeMaxVisibleHeight } =
    normalizeNodeVisibleMinMax(params);
  let minScaleX = 0;
  let minScaleY = 0;
  let maxScaleX = Infinity;
  let maxScaleY = Infinity;
  if (minNodeWidth) {
    minScaleX = nodeMinVisibleWidth / minNodeWidth;
    maxScaleX = nodeMaxVisibleWidth / minNodeWidth;
  }
  if (minNodeHeight) {
    minScaleY = nodeMinVisibleHeight / minNodeHeight;
    maxScaleY = nodeMaxVisibleHeight / minNodeHeight;
  }
  return {
    minScaleX,
    minScaleY,
    maxScaleX,
    maxScaleY,
  };
};

export const getZoomFitMatrix = (args: IZoomFitParams): ITransformMatrix => {
  const { data, graphConfig, disablePan, direction, rect } = args;

  const { nodes } = data;

  // istanbul ignore next
  if (nodes.size === 0) {
    return [1, 0, 0, 1, 0, 0];
  }

  const { minNodeWidth, minNodeHeight, minNodeX, minNodeY, maxNodeX, maxNodeY } = getContentArea(data, graphConfig);

  const { minScaleX, minScaleY, maxScaleX, maxScaleY } = getScaleRange(args, {
    width: minNodeWidth,
    height: minNodeHeight,
  });

  const spacing = normalizeSpacing(args.spacing);
  const { width, height } = rect;
  const scaleX = width / (maxNodeX - minNodeX + spacing.left + spacing.right);
  const scaleY = height / (maxNodeY - minNodeY + spacing.top + spacing.bottom);

  const scaleCommon =
    direction === Direction.Y
      ? Math.min(Math.max(minScaleX, minScaleY, scaleY), maxScaleX, maxScaleY)
      : Math.min(Math.max(minScaleX, minScaleY, Math.min(scaleX, scaleY)), maxScaleY, maxScaleY);

  const newScaleX = direction === Direction.XY ? Math.min(Math.max(minScaleX, scaleX), maxScaleX) : scaleCommon;
  const newScaleY = direction === Direction.XY ? Math.min(Math.max(minScaleY, scaleY), maxScaleY) : scaleCommon;

  if (disablePan) {
    return [newScaleX, 0, 0, newScaleY, 0, 0];
  }

  const dx = -newScaleX * (minNodeX - spacing.left);
  const dy = -newScaleY * (minNodeY - spacing.top);

  const visibleNodes = getVisibleNodes(
    data.nodes,
    {
      rect,
      transformMatrix: [newScaleX, 0, 0, newScaleY, dx, dy],
    },
    graphConfig,
  );

  if (visibleNodes.length > 0) {
    return [newScaleX, 0, 0, newScaleY, dx, dy];
  }

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  let focusNode = data.nodes.first()!;
  if (focusNode) {
    data.nodes.forEach(node => {
      if (focusNode.y > node.y) {
        focusNode = node;
      }
    });
  }

  return [
    newScaleX,
    0,
    0,
    newScaleY,
    -newScaleX * (focusNode.x - spacing.left),
    -newScaleY * (focusNode.y - spacing.top),
  ];
};

export const focusArea = (
  minX: number,
  minY: number,
  maxX: number,
  maxY: number,
  viewport: Required<IViewport>,
): IViewport => {
  const width = maxX - minX;
  const height = maxY - minY;

  const scale = Math.min(viewport.rect.width / width, viewport.rect.height / height);

  const dx = -scale * (minX + width / 2) + viewport.rect.width / 2;
  const dy = -scale * (minY + height / 2) + viewport.rect.height / 2;

  return {
    ...viewport,
    transformMatrix: [scale, 0, 0, scale, dx, dy],
  };
};
export const getContainer = (svgRef: React.RefObject<SVGSVGElement>) =>
  svgRef.current?.closest<HTMLDivElement>(".react-dag-editor-container");

/**
 * @param rectRef
 */
export function getContainerCenter(rectRef: React.RefObject<IContainerRect | undefined>): IPoint | undefined {
  const rect = rectRef.current;
  if (!rect) {
    return undefined;
  }
  const x = rect.width / 2;
  const y = rect.height / 2;
  return {
    x,
    y,
  };
}

/**
 * @param rect
 * @param e
 */
export function getRelativePoint(rect: IContainerRect, e: MouseEvent): IPoint {
  const x = e.clientX - rect.left;
  const y = e.clientY - rect.top;
  return {
    x,
    y,
  };
}

// scroll the (x,y) to the center of the viewport, or to the specific position if provide
export const scrollIntoView = (
  x: number,
  y: number,
  rect: IContainerRect | undefined | null,
  alwaysScroll?: boolean,
  position?: IPoint,
): Action => {
  // istanbul ignore next
  if (!rect) {
    return identical;
  }
  const { width, height } = rect;
  const shouldScroll = x < 0 || x > width || y < 0 || y > height;

  if (!shouldScroll && !alwaysScroll) {
    return identical;
  }

  return prevState => {
    const dx = position ? position.x - x : width / 2 - x;
    const dy = position ? position.y - y : height / 2 - y;

    return {
      ...prevState,
      transformMatrix: [
        prevState.transformMatrix[0],
        prevState.transformMatrix[1],
        prevState.transformMatrix[2],
        prevState.transformMatrix[3],
        prevState.transformMatrix[4] + dx,
        prevState.transformMatrix[5] + dy,
      ],
    };
  };
};

export const getScaleLimit = (data: GraphModel, settings: IGraphSettings) => {
  const { minNodeWidth, minNodeHeight } = getContentArea(data, settings.graphConfig);

  const { minScaleX, minScaleY } = getScaleRange(settings, {
    width: minNodeWidth,
    height: minNodeHeight,
  });

  return Math.max(minScaleX, minScaleY);
};
