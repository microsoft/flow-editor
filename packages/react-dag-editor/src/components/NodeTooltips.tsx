import * as React from "react";
import { useGraphConfig } from "../hooks/context";
import type { IViewport } from "../models/geometry";
import type { NodeModel } from "../models/NodeModel";
import { GraphNodeStatus } from "../models/status";
import { getNodeConfig } from "../utils";
import * as Bitset from "../utils/bitset";

interface IProps {
  node?: NodeModel;
  viewport: Required<IViewport>;
}

export const NodeTooltips: React.FunctionComponent<IProps> = (props) => {
  const { node, viewport } = props;
  const graphConfig = useGraphConfig();

  if (!node) {
    return null;
  }

  if (!Bitset.has(GraphNodeStatus.Activated)(node.status)) {
    return null;
  }

  const nodeConfig = getNodeConfig(node, graphConfig);

  if (!nodeConfig?.renderTooltips) {
    return null;
  }

  return (
    <div className="node-tooltips">
      {nodeConfig.renderTooltips({
        model: node,
        viewport,
      })}
    </div>
  );
};
