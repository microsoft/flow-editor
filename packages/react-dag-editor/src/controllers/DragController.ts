/* eslint-disable no-invalid-this */
import { IEventProvider, IGlobalMoveEventTypes } from "../event-provider/types";
import { IPoint } from "../models/geometry";
import { noop } from "../utils/noop";
import { animationFramed, IScheduledCallback } from "../utils/scheduling";

export type TGetPositionFromEvent = (e: MouseEvent) => IPoint;

export interface IOnDragMove<T = MouseEvent> {
  /**
   * clientX after getPositionFromEvent
   */
  clientX: number;
  /**
   * clientY after getPositionFromEvent
   */
  clientY: number;
  dx: number;
  dy: number;
  totalDX: number;
  totalDY: number;
  e: T;
}

export interface IOnDragEnd<T = MouseEvent> {
  totalDX: number;
  totalDY: number;
  e: T;
}

export class DragController<T extends MouseEvent = MouseEvent> {
  public onMove: (move: IOnDragMove) => void = noop;
  public onEnd: (end: IOnDragEnd) => void = noop;
  protected mouseMove: IScheduledCallback<[T]>;
  protected lastEvent: T | null = null;
  private readonly getPositionFromEvent: TGetPositionFromEvent;
  private startX = 0;
  private startY = 0;
  private prevClientX = 0;
  private prevClientY = 0;
  private readonly eventProvider: IEventProvider<IGlobalMoveEventTypes>;

  public constructor(
    eventProvider: IEventProvider<IGlobalMoveEventTypes>,
    getPositionFromEvent: TGetPositionFromEvent
  ) {
    this.eventProvider = eventProvider;
    this.getPositionFromEvent = getPositionFromEvent;
    this.mouseMove = animationFramed((e: T) => {
      this.doOnMouseMove(e);
    });
  }

  public start(e: T): void {
    this.lastEvent = e;
    const { x, y } = this.getPositionFromEvent(e);
    this.startX = x;
    this.startY = y;
    this.prevClientX = x;
    this.prevClientY = y;

    this.eventProvider.on("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.on("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  public stop(): void {
    this.mouseMove.cancel();
    this.eventProvider.off("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.off("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  protected getDelta(x: number, y: number): IPoint {
    const clientDX = x - this.prevClientX;
    const clientDY = y - this.prevClientY;
    this.prevClientX = x;
    this.prevClientY = y;
    return {
      x: clientDX,
      y: clientDY,
    };
  }

  protected getTotalDelta(e: MouseEvent): IPoint {
    const x = e.clientX - this.startX;
    const y = e.clientY - this.startY;
    return {
      x,
      y,
    };
  }

  protected doOnMouseMove(e: MouseEvent): void {
    const { x, y } = this.getPositionFromEvent(e);
    const { x: dx, y: dy } = this.getDelta(x, y);
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onMove({
      clientX: x,
      clientY: y,
      dx,
      dy,
      totalDX,
      totalDY,
      e,
    });
  }

  protected doOnMouseUp(e: T): void {
    e.preventDefault();
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onEnd({ totalDX, totalDY, e });

    this.stop();
  }

  private readonly onMouseUp = (e: T) => {
    this.lastEvent = e;
    this.doOnMouseUp(e);
    this.lastEvent = null;
  };

  private readonly onMouseMove = (e: T) => {
    this.lastEvent = e;
    e.preventDefault();
    this.mouseMove(e);
  };
}

export function defaultGetPositionFromEvent<T extends MouseEvent>(
  e: T
): IPoint {
  return {
    x: e.clientX,
    y: e.clientY,
  };
}
