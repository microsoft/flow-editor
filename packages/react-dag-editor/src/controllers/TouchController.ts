/* eslint-disable no-invalid-this */
import * as React from "react";
import { animationFramed } from "../utils/scheduling";

export interface ITouchHandler<ExtraArgs extends unknown[] = []> {
  onStart(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void;
  onMove(
    events: Map<number, PointerEvent>,
    e: PointerEvent,
    ...args: ExtraArgs
  ): void;
  onEnd(e: PointerEvent, ...args: ExtraArgs): void;
}

export class TouchController<ExtraArgs extends unknown[] = []> {
  public readonly eventHandlers = {
    onPointerDown: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers = new Map(this.pointers);
      this.pointers.set(e.pointerId, e.nativeEvent);
      this.updateHandler(e.nativeEvent, ...args);
    },

    onPointerMove: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers.set(e.pointerId, e.nativeEvent);
      this.onMove(e.nativeEvent, ...args);
    },

    onPointerUp: (e: React.PointerEvent, ...args: ExtraArgs): void => {
      if (e.pointerType !== "touch") {
        return;
      }
      e.preventDefault();
      this.pointers = new Map(this.pointers);
      this.pointers.delete(e.pointerId);
      this.updateHandler(e.nativeEvent, ...args);
    }
  };

  private pointers = new Map<number, PointerEvent>();
  private currentHandler: ITouchHandler<ExtraArgs> | undefined;
  private readonly handlers: Map<number, ITouchHandler<ExtraArgs>>;

  private readonly onMove = animationFramed(
    (e: PointerEvent, ...args: ExtraArgs) => {
      this.currentHandler?.onMove(this.pointers, e, ...args);
    }
  );

  public constructor(handlers: Map<number, ITouchHandler<ExtraArgs>>) {
    this.handlers = handlers;
  }

  private updateHandler(e: PointerEvent, ...args: ExtraArgs): void {
    const handler = this.handlers.get(this.pointers.size);
    if (handler !== this.currentHandler) {
      this.currentHandler?.onEnd(e, ...args);
      this.currentHandler = handler;
      this.currentHandler?.onStart(this.pointers, e, ...args);
    }
  }
}
