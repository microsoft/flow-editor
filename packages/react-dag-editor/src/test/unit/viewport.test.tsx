import { render } from "@testing-library/react";
import * as React from "react";
import { act } from "react-dom/test-utils";
import { Direction, GraphCanvasEvent, GraphModel, IPoint, ITransformMatrix, IViewport } from "../../index";
import { Transform } from "../../src/components/Transform";
import { GraphController } from "../../src/controllers/GraphController";
import { getRenderedEdges, getRenderedNodes } from "../../src/utils/viewport";
import { GraphControllerRef, TestComponent } from "../TestComponent";
import { getGraphConfig, mockBoundingBox } from "../utils";
import { getSample1Data } from "./__data__/getSample1Data";

describe("viewport", () => {
  const rect: DOMRect | ClientRect = {
    bottom: 816,
    height: 800,
    left: 316,
    right: 1232,
    top: 16,
    width: 916,
    x: 316,
    y: 16
  };
  const data = getSample1Data();
  const viewport = (transformMatrix: ITransformMatrix): Required<IViewport> => ({
    transformMatrix,
    rect
  });

  it("getRenderedNodes", () => {
    const renderedNodes = getRenderedNodes(GraphModel.fromJSON(data).nodes, viewport([1, 0, 0, 1, 0, 0]));
    expect(renderedNodes.length).toBe(8);

    const renderedNodes2 = getRenderedNodes(
      GraphModel.fromJSON(data).nodes,
      viewport([4.77, 0, 0, 4.77, -2281.56, -1995.55])
    );
    expect(renderedNodes2.length).toBe(5);
  });

  it("getRenderedEdges", () => {
    const { nodes, edges } = GraphModel.fromJSON(data);
    const graphConfig = getGraphConfig();

    const renderedEdges = getRenderedEdges(edges, nodes, graphConfig, viewport([1, 0, 0, 1, 0, 0]));
    expect(renderedEdges.length).toBe(8);

    const renderedEdges2 = getRenderedEdges(
      edges,
      nodes,
      graphConfig,
      viewport([4.77, 0, 0, 4.77, -2281.56, -1995.55])
    );
    expect(renderedEdges2.length).toBe(7);
  });
});

describe("test zoom", () => {
  let graphController: GraphController;
  function zoom(scale: number, direction?: Direction): void {
    return graphController.dispatch({
      type: GraphCanvasEvent.Zoom,
      scale,
      direction
    });
  }
  function zoomTo(scale: number, anchor?: IPoint | undefined, direction?: Direction): void {
    graphController.dispatch({
      type: GraphCanvasEvent.ZoomTo,
      scale,
      anchor,
      direction
    });
  }
  const getViewport = () => graphController.state.viewport;

  beforeEach(() => {
    mockBoundingBox();
    const graphControllerRef = React.createRef<GraphController>();
    render(
      <TestComponent>
        <GraphControllerRef ref={graphControllerRef} />
      </TestComponent>
    );
    graphController = graphControllerRef.current!;
    expect(graphController).toBeDefined();
  });

  it("should zoom x,y", () => {
    act(() => {
      zoom(2);
    });
    let transformMatrix = getViewport().transformMatrix;
    expect(transformMatrix[0]).toBe(2);
    expect(transformMatrix[3]).toBe(2);

    act(() => {
      zoomTo(3);
    });
    transformMatrix = getViewport().transformMatrix;
    expect(transformMatrix[0]).toBe(3);
    expect(transformMatrix[3]).toBe(3);
  });

  it("should zoom x only", () => {
    act(() => {
      zoom(2, Direction.X);
    });
    let transformMatrix = getViewport().transformMatrix;
    expect(transformMatrix[0]).toBe(2);
    expect(transformMatrix[3]).toBe(1);

    act(() => {
      zoomTo(3, undefined, Direction.X);
    });
    transformMatrix = getViewport().transformMatrix;
    expect(transformMatrix[0]).toBe(3);
    expect(transformMatrix[3]).toBe(1);
  });

  it("should zoom y only", () => {
    act(() => {
      zoom(2, Direction.Y);
    });
    let transformMatrix = getViewport().transformMatrix;
    expect(transformMatrix[0]).toBe(1);
    expect(transformMatrix[3]).toBe(2);

    act(() => {
      zoomTo(3, undefined, Direction.Y);
    });
    transformMatrix = getViewport().transformMatrix;
    expect(transformMatrix[0]).toBe(1);
    expect(transformMatrix[3]).toBe(3);
  });
});

describe("test Transform", () => {
  it("should get correct transform", () => {
    expect(render(<Transform matrix={[1, 0, 0, 1, 0, 0]} />).container).toMatchSnapshot();
  });
});
