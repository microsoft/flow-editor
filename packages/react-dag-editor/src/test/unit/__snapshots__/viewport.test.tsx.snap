// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Transform should get correct transform 1`] = `
<div>
  <g
    transform="matrix(1 0 0 1 0 0)"
  />
</div>
`;

exports[`test Transform should get correct transform: events 1`] = `[]`;

exports[`test zoom should zoom x only: events 1`] = `
[
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
]
`;

exports[`test zoom should zoom x,y: events 1`] = `
[
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
]
`;

exports[`test zoom should zoom y only: events 1`] = `
[
  "[Canvas]ViewportResize",
  "[Canvas]VirtualizationRecalculated",
]
`;

exports[`viewport getRenderedEdges: events 1`] = `[]`;

exports[`viewport getRenderedNodes: events 1`] = `[]`;
